conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +317,434 @@ def conda_http_errors(url, repodata_fn):
+        return self.get("etag", "")
+
+    @etag.setter
+    def etag(self, value):
+        self["etag"] = value or ""
+
+    @property
+    def cache_control(self) -> str:
+        """
+        Cache-Control header or ""
+        """
+        return self.get("cache_control", "")
+
+    @cache_control.setter
+    def cache_control(self, value):
+        self["cache_control"] = value or ""
+
+    def has_format(self, format: str) -> tuple[bool, datetime.datetime | None]:
+        # "has_zst": {
+        #     // UTC RFC3999 timestamp of when we last checked whether the file is available or not
+        #     // in this case the `repodata.json.zst` file
+        #     // Note: same format as conda TUF spec
+        #     // Python's time.time_ns() would be convenient?
+        #     "last_checked": "2023-01-08T11:45:44Z",
+        #     // false = unavailable, true = available
+        #     "value": BOOLEAN
+        # },
+
+        key = f"has_{format}"
+        if key not in self:
+            return (True, None)  # we want to check by default
+
+        try:
+            obj = self[key]
+            last_checked_str = obj["last_checked"]
+            if last_checked_str.endswith("Z"):
+                last_checked_str = f"{last_checked_str[:-1]}+00:00"
+            last_checked = datetime.datetime.fromisoformat(last_checked_str)
+            value = bool(obj["value"])
+            return (value, last_checked)
+        except (KeyError, ValueError, TypeError) as e:
+            log.warn("error parsing `has_` object from `<cache key>.state.json`", exc_info=e)
+            self.pop(key)
+
+        return False, datetime.datetime.now(tz=datetime.timezone.utc)
+
+    def set_has_format(self, format: str, value: bool):
+        key = f"has_{format}"
+        self[key] = {
+            "last_checked": datetime.datetime.now(tz=datetime.timezone.utc).isoformat()[
+                : -len("+00:00")
+            ]
+            + "Z",
+            "value": value,
+        }
+
+    def should_check_format(self, format: str) -> bool:
+        """
+        Return True if named format should be attempted.
+        """
+        has, when = self.has_format(format)
+        return (
+            has is True
+            or isinstance(when, datetime.datetime)
+            and datetime.datetime.now(tz=datetime.timezone.utc) - when
+            > CHECK_ALTERNATIVE_FORMAT_INTERVAL
+        )
+
+    def __setitem__(self, key: str, item: Any) -> None:
+        if key in self._aliased:
+            key = key[1:]  # strip underscore
+        return super().__setitem__(key, item)
+
+    def __missing__(self, key: str):
+        if key in self._aliased:
+            key = key[1:]  # strip underscore
+        else:
+            raise KeyError(key)
+        return super().__getitem__(key)
+
+
+LOCK_BYTE = 21  # mamba interop
+LOCK_ATTEMPTS = 10
+LOCK_SLEEP = 1
+
+
+@contextmanager
+def _lock_noop(fd):
+    """
+    When locking is not available.
+    """
+    yield
```

Comment:
```
>>> import contextlib
>>> foo = contextlib.nullcontext()
>>> with foo('bar', 'baz'):
...     print('hello')
TypeError: 'nullcontext' object is not callable
```

Reply:
that's not how you use it, you'd end up with doing the following instead of defining `_lock_noop`:

```
def _lock(fd):
    if "jlap" in context.experimental or "lock" in context.experimental:
        # locking required for jlap
        return _lock_impl(fd)
    return contextlib.nullcontext()
```

Reply:
Since we use the same function for "locking wanted but not supported on platform" it is simpler to retain `_lock_noop`